pkg/nydussdk/client.go

```diff
@@ -20,233 +20,331 @@ import (
+	url := c.url(endpointMetrics, query)
+	req, err := http.NewRequest(http.MethodGet, url, nil)
 	if err != nil {
-		return nil, errors.Wrap(err, "failed to create GetFsMetric request")
+		return nil, errors.Wrapf(err, "construct request url %s", url)
 	}
+
 	resp, err := c.httpClient.Do(req)
 	if err != nil {
-		return nil, errors.Wrapf(err, "failed to do HTTP GET to %s", getStatURL)
+		return nil, err
 	}
 	defer resp.Body.Close()
-	if resp.StatusCode == http.StatusNoContent {
-		return nil, errors.New("got unexpected http status StatusNoContent")
+
+	if !succeeded(resp) {
+		return nil, errors.New("failed to get metrics")
 	}
 
 	var m model.FsMetric
-	if err = json.NewDecoder(resp.Body).Decode(&m); err != nil {
-		return nil, errors.Wrap(err, "failed to decode FsMetric")
+	err = decode(resp, &m)
+	if err != nil {
+		return nil, err
 	}
+
 	return &m, nil
 }
 
-func (c *NydusClient) SharedMount(sharedMountPoint, bootstrap, daemonConfig string) error {
-	requestURL := fmt.Sprintf("http://unix%s?mountpoint=%s", mountEndpoint, sharedMountPoint)
-	content, err := os.ReadFile(daemonConfig)
+func (c *NydusdClient) SharedMount(mp, bootstrap, daemonConfig string) error {
+	config, err := os.ReadFile(daemonConfig)
 	if err != nil {
-		return errors.Wrapf(err, "failed to get content of daemon config %s", daemonConfig)
+		return errors.Wrapf(err, "read nydusd configurations %s", daemonConfig)
 	}
-	body, err := json.Marshal(model.NewMountRequest(bootstrap, string(content)))
+
+	body, err := json.Marshal(model.NewMountRequest(bootstrap, string(config)))
 	if err != nil {
-		return errors.Wrap(err, "failed to create mount request")
+		return errors.Wrap(err, "construct mount request")
 	}
-	resp, err := c.httpClient.Post(requestURL, contentType, bytes.NewBuffer(body))
+
+	query := query{}
+	query.set("mountpoint", mp)
+	url := c.url(endpointMount, query)
+
+	resp, err := c.httpClient.Post(url, jsonContentType, bytes.NewBuffer(body))
 	if err != nil {
-		return errors.Wrapf(err, "failed to do HTTP POST to %s", requestURL)
+		return err
 	}
 	defer resp.Body.Close()
-	if resp.StatusCode == http.StatusNoContent {
+
+	if succeeded(resp) {
 		return nil
 	}
-	return handleMountError(resp)
-}
 
-func (c NydusClient) FscacheBindBlob(daemonConfig string) error {
-	log.L.Infof("requesting daemon to bind fscache blob with config %s", daemonConfig)
+	return parseErrorMessage(resp)
+}
 
+func (c *NydusdClient) FscacheBindBlob(daemonConfig string) error {
+	// FIXME: it brings extra IO latency!
 	body, err := os.ReadFile(daemonConfig)
 	if err != nil {
-		return errors.Wrapf(err, "failed to get content of daemon config %s", daemonConfig)
+		return errors.Wrapf(err, "read daemon configuration %s", daemonConfig)
 	}
 
-	requestURL := "http://unix/api/v2/blobs"
-	req, err := http.NewRequest(http.MethodPut, requestURL, bytes.NewBuffer(body))
+	url := c.url(endpointBlobs, query{})
+	req, err := http.NewRequest(http.MethodPut, url, bytes.NewBuffer(body))
```

Comment:
Maybe use `simpleRequest` ?

Reply:
I am afraid current `simpleRequest` can't work with http body which means it is not simple request as defined in the helper

Reply:
`simpleRequest` now can receive body